protocols/messages.go

```diff
@@ -47,13 +47,12 @@ func (p messagePayload) hasProposal() bool {
 // Type returns the type of the payload, either a SignedProposal or SignedState.
 func (p messagePayload) Type() PayloadType {
 	switch {
-	case p.hasProposal() && !p.hasState():
+	case p.hasProposal():
 		return SignedProposalPayload
-	case !p.hasProposal() && p.hasState():
+	case p.hasState():
 		return SignedStatePayload
-	case p.hasProposal() && p.hasState():
-		panic("payload has both state and proposal %v")
 	default:
+		// This package guarantees that payloads have exactly one state or proposal
 		panic("payload has neither state nor proposal")
```

Comment:
I'm not sure this is a good change. We're lumping the "neither" and "both" cases together and panicking with a message of "neither". The panic "should never happen" but if it does, it would be better to have the extra information (and not something potentially misleading). 

There's a small chance the panic will be triggered in future (after a misguided refactor, say). If we were 100% sure it would never be triggered, then we wouldn't even need the default case in the switch at all.

Reply:
Are you suggesting that `Type()` should continue to check:
- if there's a signed state, there's no other data
- if there's a signed proposal, there's no other data
?

My goal is to allow a Payload to contain more types of data. If there are `n` data types, then this sort of validation would perform up to `n^2` checks in the worst case. I am not worried about performance, but it is definitely not ideal.

Reply:
I agree that the full validation here does not scale well and is not necessary if full validation is done at construction time.

Reply:
Right. In that case, `Type()` can assume that validation has been performed, which means that "both" is not possible. (Also, notice that if "both" were present, `Type()` would return `SignedPayloadType` instead of panicking.)

Given that, do you still see an issue with it?

Reply:
If I were to nitpick, then yes. By the same argument "neither" is not possible, but yet we are handling that with a panic.

It would be less confusing to me if we had:

```go
if p.hasState() {
   return SignedStatePayload
} 
return SignedProposalPayload
```

Reply:
I am happy with doing it that way, as it is functionally equivalent. On the other hand, it has the downside of returning the wrong value when a new Type is introduced, rather than panicking. This increases the chance of a mistake slipping past the test suite.

fd1e42dffbe1afc5e103a5c3bc4f732a4efc0701